### ISSUE TITLE
Ch12 back propagation - sigmoid derivative not included

### ISSUE THREAD

**Issue by jeevandev** [rasbt/python-machine-learning-book-2nd-edition#54]
[ch12 notebook](https://github.com/rasbt/python-machine-learning-book-2nd-edition/blob/master/code/ch12/ch12.ipynb)        
```
##################
# Backpropagation
##################

# [n_samples, n_classlabels]
sigma_out = a_out - y_train_enc[batch_idx]
```

isn't this supposed to be 
```
sigma_out = np.multiply(-(y_train_enc[batch_idx] - a_out), sigmoid_derivative_out)
```

Reference: https://github.com/stephencwelch/Neural-Networks-Demystified/blob/master/Part%204%20Backpropagation.ipynb

**Comment by rasbt** [comment]
Thanks for the note. However, `a_out - y_train_enc[batch_idx]` should be correct as the derivative of the cross-entropy loss with softmax activation is simply the "output-target". That's because in this particular combination

(partial_Err / partial_activation) * (partial_activation / partial_netinput), the denominators multiply out. Where activation=sigmoid in this case.

I've just sketched this out on paper (for efficiency ;) ) :

![1](https://user-images.githubusercontent.com/5618407/36293482-2a94a3bc-12a8-11e8-8ec0-d14da9414613.jpg)

I noticed in the blog post that a squared loss is used (whereas in the book, I used cross entropy), hence in my case, `a_out - y_train_enc[batch_idx]` is the the partial derivative of the loss with respect to the net input z, whereas in the blog post, it's the partial derivative of the loss with respect to the activation; hence they multiply it another time with the sigmoid derivative (in my case, it's included due to the cancellation mentioned earlier). 

In any case, I hope that helps :)
